lib/seahorn/BvOpSem2.cc

```diff
@@ -852,23 +852,25 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
       Expr underflow =
           m_ctx.alu().doBaddNoUnderflow(op0, op1, ty->getScalarSizeInBits());
       Expr is_carry = m_ctx.alu().doBNeg(
-          m_ctx.alu().doAnd(overflow, underflow, ty->getScalarSizeInBits()),
+          m_ctx.alu().doAnd(overflow, underflow, 1 /* bitwidth */),
           1 /* bitwidth*/);
       Expr carry_1bv = m_ctx.alu().boolToBv1(is_carry);
 
       const DataLayout DL = m_sem.getDataLayout();
       Type *curTy = I.getType();
       auto *STy = dyn_cast<StructType>(curTy);
       const StructLayout *SL = DL.getStructLayout(STy);
-      curTy = STy->getElementType(1);
-      auto sizeInBits = curTy->getScalarSizeInBits();
-      LOG("opsem", WARN << "Struct field1 size (in bits):" << sizeInBits << "\n");
-
-      Expr carry_zext =
-          m_ctx.alu().doZext(carry_1bv, sizeInBits, ty->getScalarSizeInBits());
-
-      Expr res = bv::concat(add_res, carry_zext);
-
+      auto struc_size_in_bits = SL->getSizeInBits();
+      auto offset_in_bits = SL->getElementOffsetInBits(1);
+      auto carry_size_in_bits = (STy->getElementType(1))->getScalarSizeInBits();
+      LOG("opsem", WARN << "Struct field1 size (in bits):" << struc_size_in_bits
+                        << " " << offset_in_bits << " " << carry_size_in_bits
+                        << "\n");
+      Expr carry =
+          bv::concat(carry_1bv, m_ctx.alu().si(0U, struc_size_in_bits - offset_in_bits -
```

Comment:
`bv::concat` can be moved to ALU. It should be also given bitwidth of each argument being concatenated. Then it can do the necessary booltobv1 coversion